gazebo/sensors/GpuRaySensor.cc

```diff
@@ -804,6 +972,69 @@ double GpuRaySensor::NextRequiredTimestamp() const
 void GpuRaySensor::ResetLastUpdateTime()
 {
   Sensor::ResetLastUpdateTime();
-  if (this->useStrictRate)
+  if (GpuRaySensor::useStrictRate)
     this->dataPtr->nextRenderingTime = std::numeric_limits<double>::quiet_NaN();
 }
+
+//////////////////////////////////////////////////
+unsigned int GpuRaySensor::CameraCount() const
+{
+  return this->dataPtr->laserCam->CameraCount();
+}
+
+//////////////////////////////////////////////////
+double GpuRaySensor::HorzHalfAngle() const
+{
+  return (this->AngleMax() + this->AngleMin()).Radian() / 2.0;
+}
+
+//////////////////////////////////////////////////
+double GpuRaySensor::VertHalfAngle() const
+{
+  return (this->VerticalAngleMax() + this->VerticalAngleMin()).Radian() / 2.0;
+}
+
+//////////////////////////////////////////////////
+bool GpuRaySensor::readCsvFile(std::string file_name, std::vector<std::vector<double>>& datas)
+{
+  std::fstream file_stream;
+  file_stream.open(file_name, std::ios::in);
+  if (file_stream.is_open()) {
+      std::string header;
+      std::getline(file_stream, header, '\n');
+      while (!file_stream.eof()) {
+          std::string line_str;
+          std::getline(file_stream, line_str, '\n');
+          std::stringstream line_stream;
+          line_stream << line_str;
+          std::vector<double> data;
+          try {
+              while (!line_stream.eof()) {
+                  std::string value;
+                  std::getline(line_stream, value, ',');
+                  data.push_back(std::stod(value));
+              }
+          } catch (...) {
+              continue;
+          }
+          datas.push_back(data);
+      }
+      gzdbg << "data size:" << datas.size() << "\n";
+      return true;
+  } else {
+      gzerr << "cannot read csv file!" << file_name << "\n";
+  }
+  return false;
+}
```

Comment:
Seems to be copied from livox plugin, maybe add notice to license on top to be on the safe side?

Reply:
Ah right, the copy of the function was more or less for testing. I replaced it now with my own function to load the csv file.